versioned_docs/version-4.0.0-alpha.1/datastructures/mapping.md

```diff
@@ -3,76 +3,91 @@ title: Working with Mapping
-        pub fn new(count: u32) -> Self {
-            // This call is required in order to correctly initialize the
-            // `Mapping`s of our contract.
-            ink_lang::utils::initialize_contract(|contract: &mut Self| {
-                let caller = Self::env().caller();
-                contract.map.insert(&caller, &count);
-            })
+        /// Constructor to initialize the contract with an empty mapping.
+        #[ink(constructor, payable)]
+        pub fn new() -> Self {
+            let balances = Mapping::default();
+            Self { balances }
         }
 
-        #[ink(constructor)]
-        pub fn default() -> Self {
-            // Even though we're not explicitly initializing the `Mapping`,
-            // we still need to call this
-            ink_lang::utils::initialize_contract(|_| {})
+        /// Retreive the balance of the caller.
+        #[ink(message)]
+        pub fn get_balance(&self) -> Option<Balance> {
+            let caller = self.env().caller();
+            self.balances.get(caller)
         }
 
-        // Grab the number at the caller's AccountID, if it exists
-        #[ink(message)]
-        pub fn get(&self) -> u32 {
-            let caller = Self::env().caller();
-            self.map.get(&caller).unwrap_or_default()
+        /// Credit more money to the contract.
+        #[ink(message, payable)]
+        pub fn transfer(&mut self) {
+            let caller = self.env().caller();
+            let balance = self.balances.get(caller).unwrap_or(0);
+            let endowment = self.env().transferred_value();
+            self.balances.insert(caller, &(balance + endowment));
+        }
+
+        /// Withdraw all your balance from the contract.
+        pub fn withdraw(&mut self) {
+            let caller = self.env().caller();
+            let balance = self.balances.get(caller).unwrap();
+            self.balances.remove(caller);
+            self.env().transfer(caller, balance).unwrap()
         }
     }
 }
+
 ```
+
+## Considerations when using the `Mapping` type
+
+### Updating values
+
+The attentive reader may have noticed that accessing mapping values via the `get()` 
+function will result in an owned value (a local copy), as opposed to a direct reference 
+into the storage. Changes to this value won't be reflected in the contracts storage 
+"automatically". To avoid this common pitfall, the value must be inserted again at the same 
+key after it was modified.
+
+### Storage loading behaviour
+
+Each mapping value lives under it's own storage key. Briefly, this means that mappings are 
+lazily loaded in ink!. In other words, if your message does only access a single key of a 
+mapping, it will not load the whole mapping but only the value being accessed.
+
+Furthermore, this implies that it is not possible to iterate over the contents of a map. 
+Circumventing this feature by storing populated keys inside a `Vec` is not adviseable 
+as this might result in very high gas costs.
```

Comment:
You should elaborate as to why, or at least point readers to a section on eager/lazy loading